src/combinatorics/simple_graph/triangle/tripartite.lean

```diff
@@ -0,0 +1,264 @@
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies, Bhavik Mehta
+-/
+import combinatorics.simple_graph.triangle.basic
+
+/-!
+# Construct a tripartite graph from its triangles
+
+This file contains the construction of a simple graph on `α ⊕ β ⊕ γ` from a list of triangles
+`(a, b, c)` (with `a` in the first summand, `b` in the second, `c` in the third).
+
+We call
+* `t : finset (α × β × γ)` the set of *triangle indices* (its elements are not triangles within the
+  graph but instead index them).
+* *explicit* a triangle of the constructed graph coming from a triangle index.
+* *accidental* a triangle of the constructed graph not coming from a triangle.
+
+The two important properties of this construction are:
+* `simple_graph.tripartite_from_triangles.explicit_disjoint`: Whether the explicit triangles are
+  edge-disjoint.
+* `simple_graph.tripartite_from_triangles.no_accidental`: Whether all triangles are explicit.
+
+This construction shows up unrelatingly twice in the theory of Roth numbers:
+* The lower bound of the Ruzsa-Szemerédi problem: From a Salem-Spencer set `s` on a commutative ring
+  `α` (in which `2` is invertible), we build a locally linear graph on `α ⊕ α ⊕ α`. The triangle
+  indices are `(x, x + a, x + 2 * a)` for `x` any element and `a ∈ s`. The explicit triangles are
+  edge-disjoint and accidental triangles correspond to arithmetic progressions of length 3 in `s`.
+* The proof of the corners theorem from the triangle removal lemma. For a subset `s` of the `n × n`
+  grid, we construct a tripartite graph whose vertices are the horizontal, vertical and diagonal
+  lines in the grid. The explicit triangles are `(h, v, d)` where `h`, `v`, `d` are horizontal,
+  vertical, diagonal lines that intersect in an element of `s`. The explicit triangles are
+  edge-disjoint. Accidental triangles correspond to non-trivial corners.
+
+# Main declarations
+
+* `simple_graph.tripartite_from_triangles.graph`: The tripartite simple graph generated by the given
+  triangle indices.
+* `simple_graph.tripartite_from_triangles.explicit_disjoint`: Predicate on triangle indices for the
+  constructed graph to have edge-disjoint explicit triangles.
+* `simple_graph.tripartite_from_triangles.no_accidental`: Predicate on triangle indices for there to
+  be no accidental triangles.
+* `simple_graph.tripartite_from_triangles.locally_linear`: If both predicates hold, the constructed
+  graph is locally linear.
+-/
+
+open finset function sum3
+
+variables {α β γ 𝕜 : Type*} [linear_ordered_field 𝕜] {t : finset (α × β × γ)} {a a' : α} {b b' : β}
+  {c c' : γ} {x : α × β × γ} {ε : 𝕜}
+
+namespace simple_graph
+namespace tripartite_from_triangles
+
+/-- The underlying relation of the tripartite-from-triangles graph. Two vertices are related iff
+there exists a triangle index containing them both. -/
+@[mk_iff] inductive rel (t : finset (α × β × γ)) : α ⊕ β ⊕ γ → α ⊕ β ⊕ γ → Prop
```

Comment:
`α × β × γ` feels like a rather strange representation of a triangle, as it isn't order-invariant. Would it be correct to use "ordered triangles" in the docstring; and do you actually end up needing this ordering?

Reply:
Those are not the triangles but the triangle indices. The triangle index `(a, b, c)` corresponds to the triangle `{in₀ a, in₁ b, in₂ c}` (which is unordered, even though the `in₀`, `in₁`, `in₂` make it morally ordered). I definitely need the order in my applications, since `α`, `β`, `γ` won't even be the same.